src/services/utilities.ts

```diff
@@ -3412,6 +3414,39 @@ export function getContextualTypeFromParent(node: Expression, checker: TypeCheck
         }
         case SyntaxKind.CaseClause:
             return getSwitchedType(parent as CaseClause, checker);
+        case SyntaxKind.ArrayLiteralExpression:

```

Comment:
I believe that the first step to improving how tuples are autocomplete would be to improve how they are discriminated. I'd start from looking into `discriminateContextualTypeByObjectMembers` and implementing a similar function, smth like `discriminateContextualTypeByElements`

Reply:
implemented this using discriminateContextualTypeByElements, it feels a lot less hacky!  There's some failing tests that I am looking into, but lmk if this seems off track in any way